README.md
<!--
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages).

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages).
-->

![CI status](https://api.codemagic.io/apps/643d42ab38ef8225c156a310/643d42ab38ef8225c156a30f/status_badge.svg)

A package that provides basic utils like Themes, a common BuildContext
extension, and a simple Error Page widget.

Great for creating lots of Flutter apps as a portfolio.

## Features

- Tired of defining themes manually? just use the default ThemeData from this
  package:

  ```dart
  import 'package:core/core.dart' as core;

  class MyApp extends StatelessWidget {
    const MyApp({super.key});

    @override
    Widget build(BuildContext context) {
      return MaterialApp(
        title: core.appName,
        theme: core.AppThemes.light,
        darkTheme: core.AppThemes.dark,
        ...
      );
    }
  }
  ```

- Need shorthand for the common use of context? Try the BuildContext extension
  from this package:

  ```dart
  import 'package:core/core.dart';

  class MyWidget extends StatelessWidget {
    const MyWidget({super.key});

    @override
    Widget build(BuildContext context) {
      // Shorthand for Theme.of(context)
      context.theme;

      // Shorthand for MediaQuery.of(context)
      context.mediaQuery;

      // Shorthand for Scaffold.of(context)
      context.scaffoldState;

      // Check "src/utils/extensions.dart" for more

      return Container();
    }
  }
  ```

- Need a simple page to display errors from HTTP responses? Try `HttpErrorPages`
  from the package:

  ```dart
  import 'package:core/core.dart';

  // Will display status code 400 with 'Bad Request' text below it.
  HttpErrorPages.client.badRequest();

  // Will display status code 500 with 'Internal Server Error' text below it.
  HttpErrorPages.server.internalServerError();

  // You can also add custom message to the page
  // (Default message is 'Sorry for the inconvenience').
  //
  // The message displayed below statusCode and name.
  HttpErrorPages.client.notFound(message: 'The Page You Requested Not Found');

  // Create `HttpErrorPage` directly if you need a page based on the response
  // status code.
  HttpErrorPage(404); // Will return 'Not Found' page.

  // See preview below for the page preview.
  ```

  ### HttpErrorPages previews:

  ![HttpErrorPage preview](readme_assets/http_error_page_preview.png)

  - Or do you just need an `Exception` model to handle HTTP response errors? Try
    using `HttpResponseException` from this package.

    ```dart
    throw HttpResponseException(response.code);
    ```

## Getting started

1. Add the package to your dependencies:

   ```yaml
   dependencies:
     core:
       git:
         url: https://github.com/KeidsID/flutter_app_core_package.git
   ```

2. Then, you just need to call `init()` once in `main()`:

   ```dart
   import 'package:core/core.dart' as core;

   void main() {
     core.init();
     ...
   }
   ```

   You can also customize the Theme via `init()`:

   ```dart
   import 'package:core/core.dart' as core;
   import 'package:flutter/material.dart';

   void main() {
     core.init(
       appName: 'My App Name',
       appPrimaryColor: Colors.amber,
       useMaterial3: true,
     );
     ...
   }
   ```

3. You can also add `CoreLocalizations.delegate` to your app to localize the
   default message of `HttpErrorPage`.

   ```dart
   import 'package:core/core.dart' as core;
   import 'package:flutter/material.dart';

   class MyApp extends StatelessWidget {
     const MyApp({super.key});

     @override
     Widget build(BuildContext context) {
       return MaterialApp(
         ...
         localizationsDelegates: CoreLocalizations.localizationsDelegates,
         supportedLocales: CoreLocalizations.supportedLocales,
         ...
       );
     }
   }
   ```

Now, you are ready to use this package's utilities.

## Usage

Simple usage example:

```dart
import 'package:core/core.dart' as core;
import 'package:flutter/material.dart';

void main() {
  core.init(
    appName: 'Flutter Demo',
    useMaterial3: true,
  );

  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: core.appName, // "Flutter Demo"
      // Both theme used material 3
      theme: core.AppThemes.light,
      darkTheme: core.AppThemes.dark,
      themeMode: ThemeMode.system,
      // To localize default message of [HttpErrorPage]
      localizationsDelegates: CoreLocalizations.localizationsDelegates,
      supportedLocales: CoreLocalizations.supportedLocales,
      // Simple page to display 404 Not Found
      home: core.HttpErrorPages.client.notFound(),
    );
  }
}
```

## Contributing Localizations

If you want to contribute by adding new localizations please follow these steps:

1. [Fork the repo](https://github.com/KeidsID/flutter_app_core_package/fork).
2. Copy the `lib/l10n/l10n_id.arb` file into the `lib/l10n` folder with a new
   language code, following
   [this list of ISO 859-1 codes](https://en.wikipedia.org/wiki/List_of_ISO_639-1_codes).
   Why copy `l10n_id.arb`? So that you no longer need to delete the template in
   `l10n_en.arb` and focus on localization only.
3. Update the contents in the newly created file. Especially, please update the
   `@locale` value with the corresponding ISO code.
4. Then do the `flutter gen-l10n` command to generate your localization.
5. When you're done,
   [make a new pull request](https://github.com/KeidsID/flutter_app_core_package/pulls)

More about
[Localization in the Flutter Official Documentation](https://docs.flutter.dev/development/accessibility-and-localization/internationalization).


CHANGELOG.md
## 1.2.1

### Changed

- `ErrorPage` is deprecated. Use `HttpErrorPage` instead.
- `ErrorPages` is deprecated. Use `HttpErrorPages` instead.
- The `name` property is removed in the `HttpErrorPage`. But the error name will
  be customized based on the `statusCode` property and displayed on the page.
- `HttpErrorPages.byCode()` is deprecated. Use `HttpErrorPage` directly instead.

### Added

- `AppException` (Default implementation of `Exception`. But the message is not
  private).
- `CoreLocalizations` (Localization for this package).
- `HttpResponseException` (Data model for `HttpErrorPage`).

  So `HttpErrorPage` can now be created from `HttpResponseException`.

  ```dart
  HttpErrorPage.fromException(
    HttpResponseException(...),
  );
  ```

- `HttpErrorPage` will have a default message that varies according to the given
  status code. The messages are also localized if you include
  `CoreLocalizations.delegate` to your app.

### Fixed

- Changing the `..addAll()` method on `Map` to spread operator method.

## 1.1.0

### Added

- The common use of BuildContext as a dart extension.

  - `context.theme` (Shorthand for `Theme.of(context)`)
  - `context.mediaQuery` (Shorthand for `MediaQuery.of(context)`)
  - `context.scaffoldState` (Shorthand for `Scaffold.of(context)`)
  - Check **src/utils/extensions.dart** for more.

- Basic Pages to display error status based on MDN Web Docs.
  - `ErrorPages.client` (List of client error/400-499 status codes).
  - `ErrorPages.server` (List of server error/500-599 status codes).
  - `ErrorPages.byCode()` (Get an ErrorPage based on the provided statusCode).

### Changed

- Rename the `common` folder to `src`.

## 1.0.0

### Added

- Basic App Utils Added.

  - `appName`
  - `appPrimaryColor`
  - `appSecondaryColor`
  - `AppThemes` (Provide theme on both modes).


lib/src/errors/http_error_pages.dart
import 'package:flutter/material.dart';

import 'data/client_errors_map.dart';
import 'data/server_errors_map.dart';
import 'pages/http_error_page.dart';

part 'client_error_pages.dart';
part 'server_error_pages.dart';

typedef _GetHttpErrorPage = HttpErrorPage Function({
  Key? key,
  String? message,
  Widget? child,
});

@Deprecated('Use HttpErrorPages instead. Deprecated in v1.2.1.')
typedef ErrorPages = HttpErrorPages;

/// List of simple HTTP error pages based on MDN Web Docs
/// (https://developer.mozilla.org/en-US/docs/Web/HTTP/Status).
///
/// Use it the same way you use [Colors].
///
/// ```dart
/// HttpErrorPages.client.notFound(); // return HttpErrorPage widget
/// ```
abstract class HttpErrorPages {
  /// An object with properties based on HTTP Client Error Responses
  /// (status codes 400-499).
  ///
  /// Each property is a method to return an [HttpErrorPage] with the same
  /// parameters as creating the page with the constructor
  /// (but without the `statusCode` property).
  static final client = _ClientErrorPages(
    badRequest: _getClientErrorPageMap[400]!,
    unauthorized: _getClientErrorPageMap[401]!,
    forbidden: _getClientErrorPageMap[403]!,
    notFound: _getClientErrorPageMap[404]!,
    methodNotAllowed: _getClientErrorPageMap[405]!,
    notAcceptable: _getClientErrorPageMap[406]!,
    proxyAuthenticationRequired: _getClientErrorPageMap[407]!,
    requestTimeout: _getClientErrorPageMap[408]!,
    conflict: _getClientErrorPageMap[409]!,
    gone: _getClientErrorPageMap[410]!,
    lengthRequired: _getClientErrorPageMap[411]!,
    preconditionFailed: _getClientErrorPageMap[412]!,
    payloadTooLarge: _getClientErrorPageMap[413]!,
    uriTooLong: _getClientErrorPageMap[414]!,
    unsupportedMediaType: _getClientErrorPageMap[415]!,
    rangeNotSatisfiable: _getClientErrorPageMap[416]!,
    expectationFailed: _getClientErrorPageMap[417]!,
    imATeapot: _getClientErrorPageMap[418]!,
    misdirectedRequest: _getClientErrorPageMap[421]!,
    unprocessableContent: _getClientErrorPageMap[422]!,
    locked: _getClientErrorPageMap[423]!,
    failedDependency: _getClientErrorPageMap[424]!,
    upgradeRequired: _getClientErrorPageMap[426]!,
    preconditionRequired: _getClientErrorPageMap[428]!,
    tooManyRequests: _getClientErrorPageMap[429]!,
    requestHeaderFieldsTooLarge: _getClientErrorPageMap[431]!,
    unavailableForLegalReasons: _getClientErrorPageMap[451]!,
  );

  /// An object with properties based on HTTP Server Error Responses
  /// (status codes 500-599).
  ///
  /// Each property is a method to return an [HttpErrorPage] with the same
  /// parameters as creating the page with the constructor
  /// (but without the `statusCode` property).
  static final server = _ServerErrorPages(
    internalServerError: _getServerErrorPageMap[500]!,
    notImplemented: _getServerErrorPageMap[501]!,
    badGateway: _getServerErrorPageMap[502]!,
    serviceUnavailable: _getServerErrorPageMap[503]!,
    gatewayTimeout: _getServerErrorPageMap[504]!,
    httpVersionNotSupported: _getServerErrorPageMap[505]!,
    variantAlsoNegotiates: _getServerErrorPageMap[506]!,
    insufficientStorage: _getServerErrorPageMap[507]!,
    loopDetected: _getServerErrorPageMap[508]!,
    notExtended: _getServerErrorPageMap[510]!,
    networkAuthenticationRequired: _getServerErrorPageMap[511]!,
  );

  /// Get a [HttpErrorPage] based on the provided statusCode. If the statusCode is
  /// not found, an [RangeError] will be thrown.
  @Deprecated(
      'Use HttpErrorPage directly instead. Deprecated in v1.2.1. Will be removed in the major update')
  static HttpErrorPage byCode(
    int statusCode, {
    Key? key,
    String? message,
    Widget? child,
  }) {
    final errorPagesMap = {
      ..._getClientErrorPageMap,
      ..._getServerErrorPageMap,
    };

    if (errorPagesMap[statusCode] == null) {
      throw RangeError('Status Code Not Found');
    }

    return errorPagesMap[statusCode]!(key: key, message: message, child: child);
  }
}
